### PR TITLE
API: check_broadcast cannot return None

### PR DESCRIPTION
**Pull request by lpsinger**
### Description

`_validate_input_shapes` calls `check_broadcast` and compares its return value with `None`, but `check_broadcast` (as far as I can tell) cannot return `None`.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by pllim** [comment]
This looks like a bug. If that is the case, then it needs a change log entry. And a regression test (why didn't we see this failing CI?). But let's wait to hear back from `modeling` maintainers first.

File: astropy/modeling/core.py
Comment (WilliamJamieson) [inline astropy/modeling/core.py:1069]:
```suggestion
        except IncompatibleShapeError as err:
            raise ValueError(
                "All inputs must have identical shapes or must be scalars."
            ) from err
```
Raising the error from the original would be good for stack tracing

**Review by WilliamJamieson — CHANGES_REQUESTED** [review]
This is definitely an oversight as this is handled this way in several other places.. Thanks for finding it.

**Comment by lpsinger** [comment]
You know, I realized that `IncompatibleShapeError` is a more useful error message than `ValueError`. So I just removed the re-raise entirely. This is also consistent with the old behavior, whereas re-raising it as a ValueError would have actually been an API change (since the `is None` code path was never followed, and `check_broadcast` simply would have raised).

**Review by WilliamJamieson — CHANGES_REQUESTED** [review]
This test: https://github.com/astropy/astropy/blob/dce55cd3250e66e80f3fa6b0fbf4ecb58d1c287c/astropy/modeling/tests/test_core.py#L984-L1004 needs to be refactored to:

```python
def test__validate_input_shapes():
    model = models.Gaussian1D()
    model._n_models = 2


    # Full success
    inputs = [np.array([[1, 2], [3, 4]]), np.array([[5, 6], [7, 8]])]
    assert (2, 2) == model._validate_input_shapes(inputs, model.inputs, 1)

    # Fail check_broadcast
    MESSAGE = r"All inputs must have identical shapes or must be scalars"

    # Fails because the input shape of the second input has one more axis which
    # for which the first input can be broadcasted to
    inputs = [np.array([[1, 2], [3, 4]]), np.array([[5, 6], [7, 8], [9, 10]])]
    with pytest.raises(ValueError, match=MESSAGE):
        model._validate_input_shapes(inputs, model.inputs, 1)
```

(this will need to be updated to catch the `IncompatibleShapeError`)

**Comment by WilliamJamieson** [comment]
> You know, I realized that `IncompatibleShapeError` is a more useful error message than `ValueError`. So I just removed the re-raise entirely. This is also consistent with the old behavior, whereas re-raising it as a ValueError would have actually been an API change (since the `is None` code path was never followed, and `check_broadcast` simply would have raised).
> 
> By the above reasoning, this change has no user-visible impact. So I'm going to change this to MNT.

I agree.

File: astropy/modeling/core.py
Comment (WilliamJamieson) [inline astropy/modeling/core.py:1064]: After examining the `IncompatableShapeError` I think I like previous error message better.

For  example the `test_core` test which was refactored gives the message:
```
astropy.utils.shapes.IncompatibleShapeError: ((2, 2), 0, (3, 2), 1)
```
which is a bit opaque to as to what the issue is.

**Review by WilliamJamieson — CHANGES_REQUESTED** [review]

File: astropy/modeling/tests/test_core.py
Comment (pllim) [inline astropy/modeling/tests/test_core.py:995]: Changing from ValueError to IncompatibleShapeError is a behavior change too. I don't think we should backport it if you want to keep this change, though from the comment above, I think this is going to be reverted?

**Comment by pllim** [comment]
Re: labels -- Ah, you mean the change log? If you are changing the exception type that gets raised, it still needs a change log. And as stated above, I don't think we should backport that because that is strictly not backward compatible.

File: astropy/modeling/tests/test_core.py
Comment (lpsinger) [reply astropy/modeling/tests/test_core.py:995]: Right, but the point is we never _did_ raise a ValueError here, because the code that would have re-raised the error was broken (since it was checking for the value None, which check_broadcast never returns).

Don't be fooled by this change to the unit test: it never raised a ValueError. The unit test made it _look like_ it did, because it mocked up the underlying function and the mock did not match the behavior of the real function. So the unit test was wrong too.

**Comment by lpsinger** [comment]
> > would you please update the labels?
> 
> Can you please be more specific on which label needs updating? Thanks!

I mean, remove the Bug label. This should be treated as a MNT-only change. So, no need for a changelog entry.

**Comment by pllim** [comment]
> no need for a changelog entry

Depends. Are you really changing the exception type or not?

File: astropy/modeling/tests/test_core.py
Comment (pllim) [reply astropy/modeling/tests/test_core.py:995]: Okay. Lemme see for myself what the test does now on main. Maybe then I will understand.

**Comment by lpsinger** [comment]
> > I mean, remove the Bug label
> 
> So no backport, right? I also will move the milestone to v6 then, ok?

Correct, no backport, and v6 milestone is fine.

File: astropy/modeling/tests/test_core.py
Comment (pllim) [reply astropy/modeling/tests/test_core.py:995]: Okay, yes, you are right, sorry. But what about https://github.com/astropy/astropy/pull/15209/files#r1300237995 ? Does that have to be addressed?

**Comment by lpsinger** [comment]
> > no need for a changelog entry
> 
> Depends. Are you really changing the exception type or not?

I am not changing the exception type. We currently raise an IncompatibleShapeError, and we will _still_ raise an IncompatibleShapeError. (Read https://github.com/astropy/astropy/pull/15209#discussion_r1300376238 carefully.)

File: astropy/modeling/core.py
Comment (lpsinger) [reply astropy/modeling/core.py:1064]: Alright. How about if I raise a new IncompatibleShapeError but with a better explanation?

Comment (WilliamJamieson) [reply astropy/modeling/core.py:1064]: The current message is actually pretty clear about what users should be doing.

Comment (lpsinger) [reply astropy/modeling/core.py:1064]: OK, got it. That will make this an API change, though, since `IncompatibleShapeError` doesn't support custom error messages, but `ValueError` does. Sorry, @pllim!

Comment (pllim) [reply astropy/modeling/core.py:1064]: No worries. Let me know when to update the labels again. 😆  API change needs a change log for sure.

File: astropy/modeling/tests/test_core.py
Comment (lpsinger) [reply astropy/modeling/tests/test_core.py]: I'm not sure what that link points to now.

Comment (pllim) [reply astropy/modeling/tests/test_core.py]: That was @WilliamJamieson 's comment on exception message.

File: astropy/modeling/tests/test_models.py
Comment (WilliamJamieson) [reply astropy/modeling/tests/test_models.py:127]: There should be a match on the raises.

Comment (lpsinger) [reply astropy/modeling/tests/test_models.py]: Done.

**Review by pllim — COMMENTED** [review]
LGTM now, thanks! But I'll let William approve.

**Review by WilliamJamieson — APPROVED** [review]

**Pull request merged**
